### PR TITLE
fix(compiler): invalid extraction text

### LINKED ISSUE #39195: i18n: localize message extraction producing incorrect `equiv-text` values
(See https://github.com/angular/angular/issues/38711#issuecomment-706079727 for original report)

Extraction of i18n messages from the following template HTML:

```html
<ng-container i18n="Update @@update"> <b>{{value1}}:</b> current {{value2}} to new {{value3}} </ng-container>
```

**Without ivy:** (ctype attributes removed for clarity)
```xml
<source>
  <x id="START_BOLD_TEXT" equiv-text="&lt;b&gt;"/>
  <x id="INTERPOLATION" equiv-text="{{value1}}"/>:
  <x id="CLOSE_BOLD_TEXT" equiv-text="&lt;/b&gt;"/> current
  <x id="INTERPOLATION_1" equiv-text="{{value2}}"/> to new
  <x id="INTERPOLATION_2" equiv-text="{{value3}}"/>
</source>
```

**With ivy:**
```xml
<source>
  <x id="START_BOLD_TEXT" equiv-text="&lt;b&gt;{{value1}}"/>
  <x id="INTERPOLATION" equiv-text="{{value1}}"/>:
  <x id="CLOSE_BOLD_TEXT" equiv-text="&lt;/b&gt; "/> current
  <x id="INTERPOLATION_1" equiv-text="{value2}} "/> to new
  <x id="INTERPOLATION_2" equiv-text="{value3}} "/>
</source>
```

It seems that there is some offsetting going on, which is making the start and end points in the equiv-text appear in the wrong place. I'll need to debug into this but since these equiv-text attributes are primarily to aid the translator's job, it will not be high on my priority list. Hopefully I can take a look at it later next week or the week after.

### PR DESCRIPTION
**Pull request by petebacondarwin**
Fixes #39195

### CODE REVIEW

File: packages/compiler/src/parse_util.ts [hunk 1]
```diff
@@ -101,7 +101,8 @@ export class ParseSourceFile {
 
 export class ParseSourceSpan {
   constructor(
-      public start: ParseLocation, public end: ParseLocation, public details: string|null = null) {}
+      public start: ParseLocation, public end: ParseLocation,
+      public fullStart: ParseLocation = start, public details: string|null = null) {}
```
Comment (AndrewKushnir) [inline packages/compiler/src/parse_util.ts:105]: I really like the commit description which gives a lot of context on this change! Can we may be capture it as a comment here as well (as docs to the `ParseSourceSpan` class constructor)?

**Review by AndrewKushnir — APPROVED** [review]
Thanks for the fix @petebacondarwin, it looks good (just left 1 minor comment) 👍 

I think it'd be great if @alxhub or @atscott can have a quick look too, since they worked with the source spans recently and may have more context.

Thank you.

**Comment by atscott** [comment]
Let's make it a party :) also adding @JoostK since he has done specific work on ICU source spans in #39072

File: packages/compiler/src/ml_parser/lexer.ts [hunk 2]
```diff
@@ -917,19 +917,19 @@ class PlainCharacterCursor implements CharacterCursor {
 
   getSpan(start?: this, leadingTriviaCodePoints?: number[]): ParseSourceSpan {
     start = start || this;
-    let cloned = false;
+    let fullStart: this|undefined;
```
Comment (atscott) [inline packages/compiler/src/ml_parser/lexer.ts:920]: nit: the compiler code generally prefers `null` over `undefined`. While the older code might not do that, it's at least what we _try_ to stick to when writing new code.

File: packages/compiler/src/ml_parser/lexer.ts [hunk 3]
```diff
@@ -917,19 +917,19 @@ class PlainCharacterCursor implements CharacterCursor {
...
-          cloned = true;
         }
         start.advance();
       }
     }
     return new ParseSourceSpan(
-        new ParseLocation(start.file, start.state.offset, start.state.line, start.state.column),
-        new ParseLocation(this.file, this.state.offset, this.state.line, this.state.column));
+        this.locationFromCursor(start), this.locationFromCursor(this),
+        fullStart && this.locationFromCursor(fullStart));
```
Comment (atscott) [inline packages/compiler/src/ml_parser/lexer.ts:932]: nit: I think another general style preference in new compiler code is to be explicit about the checks, making this
`fullStart !== null ? this.locationFromCursor(fullStart) : this.locationFromCursor(start)` (extracting a variable for `start` would make this a tad less verbose.

**Review by atscott — APPROVED** [review]
LGTM, only some optional style nits. I don't have any more context than @AndrewKushnir  would have on i18n source spans though :)

File: packages/compiler/src/ml_parser/lexer.ts (hunk 2 above)
Comment (petebacondarwin) [reply packages/compiler/src/ml_parser/lexer.ts:920]: Yeah - I'm on board with that and I thought about it. But since I didn't want to create a new `fullStart` location if it was the same as `start`, I was trying to utilize the `undefined` so that it would be passed through to the `ParseSourceSpan`  constructor.

I actually have worked out a way to avoid the `undefined`/`null` value altogether!

**Comment by petebacondarwin** [comment]
Thanks for the reviews @AndrewKushnir and @atscott !

File: packages/compiler/src/compiler_util/expression_converter.ts [hunk 4]
```diff
@@ -917,7 +917,8 @@ class _AstToIrVisitor implements cdAst.AstVisitor {
     if (this.baseSourceSpan) {
       const start = this.baseSourceSpan.start.moveBy(span.start);
       const end = this.baseSourceSpan.start.moveBy(span.end);
-      return new ParseSourceSpan(start, end);
+      const fullStart = this.baseSourceSpan.fullStart.moveBy(span.start);
```
Comment (JoostK) [inline packages/compiler/src/compiler_util/expression_converter.ts:920]: Is this correct? The `baseSourceSpan.fullStart` corresponds with the leading trivia of the base node (HTML AST), whereas `span` is a relative offset for an expression AST node. That expression node has leading trivia of its own, so its full start cannot be derived from the base source span's full start.

For instance:

```html
<div>  {{expr}}  </div>
     ^ fullStart = 6
       ^ start = 8
         ^ span.start = 2
```

Here, both start and full start of `expr` would be 10.

An accurate full start for the expression ast would require that `cdAst.ParseSpan` also tracks its own full start offset.

File: packages/compiler/src/template_parser/binding_parser.ts [hunk 5]
```diff
@@ -585,5 +585,7 @@ function moveParseSourceSpan(
   // The difference of two absolute offsets provide the relative offset
   const startDiff = absoluteSpan.start - sourceSpan.start.offset;
   const endDiff = absoluteSpan.end - sourceSpan.end.offset;
-  return new ParseSourceSpan(sourceSpan.start.moveBy(startDiff), sourceSpan.end.moveBy(endDiff));
+  return new ParseSourceSpan(
+      sourceSpan.start.moveBy(startDiff), sourceSpan.end.moveBy(endDiff),
+      sourceSpan.fullStart.moveBy(startDiff), sourceSpan.details);
```
Comment (JoostK) [inline packages/compiler/src/template_parser/binding_parser.ts:590]: Same here.

File: packages/compiler/test/render3/view/i18n_spec.ts [hunk 6]
```diff
@@ -509,6 +510,26 @@ describe('serializeI18nMessageForLocalize', () => {
        expect(humanizeSourceSpan(placeHolders[2].sourceSpan)).toEqual('"</b>" (22-26)');
      });
 
+  it('should create the correct placeholder source-spans when there is skipped leading whitespace',
```
Comment (JoostK) [inline packages/compiler/test/render3/view/i18n_spec.ts:513]: This test appears to be passing even when I revert the change to `getOffsetSourceSpan`

**Review by JoostK — CHANGES_REQUESTED** [review]

File: packages/compiler/src/compiler_util/expression_converter.ts (hunk 4 above)
Comment (petebacondarwin) [reply packages/compiler/src/compiler_util/expression_converter.ts:920]: I'm on my phone, so not able to fully verify.
But the point is that these ParseSpan values are computed from the parsed value of the token, which includes leading trivia, and not the source span text.
So I believe that in your example `span.start` is actually 10 not 2.

File: packages/compiler/test/render3/view/i18n_spec.ts (hunk 6 above)
Comment (JoostK) [reply packages/compiler/test/render3/view/i18n_spec.ts:513]: D'oh! I was looking at the wrong test 🤦

File: packages/compiler/src/compiler_util/expression_converter.ts (hunk 4 above)
Comment (JoostK) [reply packages/compiler/src/compiler_util/expression_converter.ts:920]: Ah yes, my interpretation of the relative `span` was indeed incorrect; it's relative to the start of the full text content, not the interpolation start. All fine then!

**Review by JoostK — APPROVED** [review]

**Comment by AndrewKushnir** [comment]
[Presubmit](http://test/OCL:340128253:BASE:340128263:1604262229785:e3453681).

**Comment by AndrewKushnir** [comment]
@petebacondarwin FYI presubmit is successful (I'm removing the "presubmit" label).

**Review by alxhub — APPROVED** [review]
